### DISCUSSION TITLE
Pixel editing aspect ratio matching printer pixel sizes

### DISCUSSION THREAD

**Discussion by research-e** [sn4k3/UVtools#913] [Ideas]
When pixel editing all the pixels are shown as square. It would be nice to have the pixels match the aspect ratio of the printer as many printers have non-square pixels (e.g. the anycubic m5s pro has 16.8x24.7um pixels), so having the pixel editing screen match that aspect ratio would give a more true representation of the print file.

Would something like this be possible if it isn't a feature already?

**Comment by sn4k3** [comment]
A pixel is a pixel and by default it doesn't represent a size but a dot/coordinate in a grid.
Images are bitmaps, formed by pixels grid which are simply virtual X/Y coordinates, you "can't" have pixels fractions.
If you open same bitmap in different display ratios, you will notice that the same bitmap will show differently across different devices, because bitmaps are "size-less".
Even if we try to compensate the ratio you will never get the accurate size because the loss of fractions.
Read more here: https://github.com/sn4k3/UVtools/discussions/866#discussioncomment-9098556

There are technics to make this feasible but the extra computing power and performance hit to manipulate large bitmaps is not worth the visual candy.

**Comment by jamesoleg** [comment]
Hey research-e,

That's an interesting idea! While most pixel art editing software displays pixels as squares, you're right that some printers have non-square pixel sizes.

While directly matching the printer's pixel aspect ratio in the editing software might not be a common feature yet, there are a few workarounds you can consider:

DPI (Dots Per Inch): Most pixel art editing software allows you to define the document size in DPI (Dots Per Inch). By calculating the desired print size and your printer's DPI, you can set the document size accordingly. This ensures the correct number of pixels translates to the intended print size.
Preview at Print Size: Some software allows you to preview your artwork at the intended print size. This can give you a good idea of how the final product will look despite the square pixel representation in the editing window.
Reference Grid: You can create a custom grid overlay within your software that matches your printer's pixel aspect ratio. This can help you visualize the final printed output while editing.
Here are some additional resources that might be helpful:

Pixel Art Tutorials: Many online resources offer tutorials on creating pixel art for specific printers. These tutorials might give you printer-specific tips and workarounds.
Pixel Art Asset Packs: If you're looking for pre-made pixel art assets designed for specific print sizes, check out some of the great packs available at the Next Level Gaming Store: [Pixel Art Assets](https://nextlevelgamingstore.net/pixel-art-assets/).
Hopefully this helps!

**Reply by research-e** [reply]
Some very neat ideas! I might have to try them to see if i can get better results

I ended up coding a quick python script which essentially watches the pixel editing screen and converts and displays the aspect ratio to match that of the printer in another window. Its not 100% perfect but it gives a very close returned image to what would be printed which is useful for sense checking what im making. I also have seen someone else use OBS studio to essentially do the same thing but it seemed a bit overkill and more compute than it should be.

**Reply by sn4k3** [reply]
If is just for checking the image you can use the tool - resize and compensate one axis, that will show you a very identical image to what would be printed.

Example (Note the screw holes which should be a circle):
![image](https://github.com/user-attachments/assets/c299cae1-af9c-49da-b861-b9a8ffab5f68)
![image](https://github.com/user-attachments/assets/901e4b5b-c253-4c27-94e0-84080570b07a)

**Reply by research-e** [reply]
This is a useful tip, thank you! the issue is with this is for features that are say 2 pixels large do not translate well whereas what im trying to achieve is a stretching of the pixels to match their exact aspect ratio. For what im doing its critical i know exactly what each pixel is doing so this aspect ratio adjustment of the bitmap is very useful.

Example using resize
![image](https://github.com/user-attachments/assets/a8da4773-9301-441a-b2e0-dba92107ca23)
![image](https://github.com/user-attachments/assets/5586de97-01d2-48fc-b98b-81cd54b7acde)

Example of stretching aspect Ratio:
![image](https://github.com/user-attachments/assets/2fea47a8-d737-4f74-a041-871acfac9b38)
